src/main/utility/status_bar.rs

```diff
@@ -118,6 +114,79 @@ impl<T: 'static + StatusBarState> StatusBar<T> {
     }
 }
 
+pub struct StatusPrinter<T: StatusBarState> {
+    state: Arc<RwLock<T>>,
+    stop_cond: Arc<(Mutex<bool>, Condvar)>,
+    thread: Option<std::thread::JoinHandle<()>>,
+}
+
+impl<T: 'static + StatusBarState> StatusPrinter<T> {
+    /// Create and start drawing the status bar.
+    pub fn new(state: T) -> Self {
+        let state = Arc::new(RwLock::new(state));
+        let stop_cond = Arc::new((Mutex::new(false), Condvar::new()));
+
+        Self {
+            state: Arc::clone(&state),
+            stop_cond: Arc::clone(&stop_cond),
+            thread: Some(std::thread::spawn(move || {
+                Self::print_loop(state, stop_cond);
+            })),
+        }
+    }
+
+    fn print_loop(state: Arc<RwLock<T>>, stop_cond: Arc<(Mutex<bool>, Condvar)>) {
+        let mut print_interval = Duration::from_secs(30);
+
+        loop {
+            {
+                let stop_guard = stop_cond.0.lock().unwrap();
+                // the condvar should check the condition fn before waiting for the notification, so
+                // we don't need to add our own preliminary stop check
+                let (stop_guard, _) = stop_cond
+                    .1
+                    .wait_timeout_while(stop_guard, print_interval, |stop| !*stop)
+                    .unwrap();
+                // either we were notified or we timed out
+                if *stop_guard {
+                    break;
+                }
+            }
+
+            // We want to write everything in as few write() syscalls as possible. Note that
+            // if we were to use eprint! with a format string like "{}{}", eprint! would
+            // always make at least two write() syscalls, which we wouldn't want.
+            std::io::stderr()
+                .write_all(format!("Progress: {}\n", *state.read().unwrap()).as_bytes())
+                .unwrap();
+            let _ = std::io::stderr().flush();
+
+            // interval is limited to 2 hours
+            const MAX_INTERVAL: Duration = Duration::from_secs(2 * 60 * 60);
+
+            // exponential backoff so that long simulations don't flood stderr
+            print_interval = std::cmp::min(print_interval.saturating_mul(2), MAX_INTERVAL);
```

Comment:
This is clever, but potentially confusing. If the last update took an hour, and it's been an hour and a half since that one, I might reasonably think the simulation is stuck.

Maybe the progress line could include something like "next update in {}"?

Reply:
I'd also lean towards an upper limit of more like 30m... maybe even 10m.

Reply:
I didn't think of that. I added a "next update" message for now, and if this ends up being confusing we can just stick with a constant interval or something.

> I'd also lean towards an upper limit of more like 30m... maybe even 10m.

I didn't want to set the interval too low since it would generate a lot of messages when people run long simulations (for example 14 day sims). I've changed this to 30 minutes for now.